docs/site/Context.md

```diff
@@ -229,3 +229,101 @@ class HelloController {
 These "sugar" decorators allow you to quickly build up your application without
 having to code up all the additional logic by simply giving LoopBack hints (in
 the form of metadata) to your intent.
+
+## Context watcher
+
+Bindings in a context can come and go. It's often desirable for an artifact
+(especially an extension point) to keep track of other artifacts (extensions).
+For example, the `RestServer` needs to know routes contributed by `controller`
+classes. Ideally, a controller can be added context even after the application
+starts.
+
+To support the dynamic tracking of such artifacts registered within a context
+chain, we introduce the `ContextWatcher` class that can be used to watch a list
+of bindings matching certain criteria depicted by a `BindingFilter` function.
+
+```ts
+import {Context, ContextWatcher} from '@loopback/context';
+
+// Set up a context chain
+const appCtx = new Context('app');
+const serverCtx = new Context(appCtx, 'server'); // server -> app
+
+// Define a binding filter to select bindings with tag `controller`
+const controllerFilter = binding => binding.tagMap.controller != null;
+
+// Watch for bindings with tag `controller`
+const watcher = serverCtx.watch(controllerFilter);
+
+// No controllers yet
+await watcher.values(); => []
+
+// Bind Controller1 to server context
+serverCtx
+  .bind('controllers.Controller1')
+  .toClass(Controller1)
+  .tag('controller');
+
+// Resolve to an instance of Controller1
+await watcher.values(); // => [an instance of Controller1];
+
+// Bind Controller2 to app context
+appCtx
+  .bind('controllers.Controller2')
+  .toClass(Controller2)
+  .tag('controller');
+
+// Resolve to an instance of Controller1 and an instance of Controller2
+await watcher.values(); // => [an instance of Controller1, an instance of Controller2];
```

Comment:
This example doesn't seem very useful to me. One can use `appCtx.find()` (perhaps with `appCtx.get()`) to get the same list of controllers. Adding all the complexity of ContextWatcher seems like overengineering in this particular use case. What benefits do you see?

Also I don't see how this can be leveraged to update the routing table when a controller was added. To do so, we need a push model where the context tells us about the values, while the example here is showing a pull model where the consumer has to periodically check.

Reply:
> This example doesn't seem very useful to me. One can use appCtx.find() (perhaps with appCtx.get()) to get the same list of controllers. Adding all the complexity of ContextWatcher seems like overengineering in this particular use case. What benefits do you see?

The key benefit of `ContextWatcher` is that it caches resolved values until bindings matching the filter function are added/removed. For most cases, we don't have to pay the penalty to find/resolve per request.

The consumer does not have to pull. The pull only happens if the cache is invalidated. The context push information into the watcher, which in turn invalidates the cache at the moment as the simplest implementation. The only downside is that we expose a `Getter` function such as `values()` to make sure the cache is up to date.

Reply:
It's also possible that we force a reload during `reset()`. But it's less optimal than waiting until `values()` is called.